### PR TITLE
[ownership] Convert ValueOwnershipKind to have an Invalid state instead of using optional.

### PR DESCRIPTION
**Pull request by gottesmm**
At Andy's request. If it creates too much noise in covered switches, we may go
back to the original.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SIL/IR/SILValue.cpp [hunk 1]
```diff
@@ -215,21 +217,27 @@ llvm::raw_ostream &swift::operator<<(llvm::raw_ostream &os,
...
-    return ValueOwnershipKind(RHSVal);
-  }
-  // Any merges with anything.
-  if (RHSVal == ValueOwnershipKind::None) {
-    return ValueOwnershipKind(LHSVal);
-  }
+  // If either lhs or rhs are invalid, return invalid.
+  if (lhsVal == ValueOwnershipKind::Invalid ||
+      rhsVal == ValueOwnershipKind::Invalid)
+    return ValueOwnershipKind::Invalid;
```
Comment (atrick) [inline lib/SIL/IR/SILValue.cpp:227]: You shouldn't be special-casing invalid here.

**Review by atrick — APPROVED** [review]
Please get rid of the special-casing for Invalid. Otherwise this looks great.

Optional is appropriate when a particular API may fail. But here "Invalid" is actually an expected state, and normal state for constraints. It just means "any" or "unknown" ownership which happens to be invalid when applied to values because we have strict ownership rules for values.

**Comment by gottesmm** [comment]
Going to fix Andys comments in a follow up.

**Pull request merged**